Adding - facebook-oss workload

ajaygk95 opened the pull request:
https://github.com/facebookarchive/oss-performance

ajaygk95 commented:
Hello,

1 - cloudsuite/debian:base-os is debian "buster" image. We have used a hhvm-3.30 (LTS) which is not present for buster (https://dl.hhvm.com/debian/dists/). Due to this I had kept ubuntu-16.04 as the base image.
I will try building hhvm-3.30 from source on buster and if its successful will update the PR with the changes.

2 - Will add some more information in readme

3 - we tried building hhvm for ARM, we faced lot of issues and the build was not successful. Also as @VamshiMuthineni mentioned hhvm on ARM is not supported.

pradeepk5 commented:
> Hello,
> 
> 1 - cloudsuite/debian:base-os is debian "buster" image. We have used hvm-3.30 (LTS) which is not present for buster (https://dl.hhvm.com/debian/dists/). Due to this I had used ubuntu-16.04 as the base image.
> I will try building hhvm-3.30 from source on buster and if its successful will update the PR with the changes.
> 
> 2 - Will add some more information in readme
> 
> 3 - we tried building hhvm for ARM, we faced lot of issues and the build was not successful. Also as @VamshiMuthineni mentioned hhvm on ARM is not supported.

When updating PR, please consider following changes

* perf is removed in one of the commits, but there are some remaining scripts which uses perf (perf.sh)
* seige load tester used in Django too, Can we keep move `seige` to common?
* Readme metions about changing parameters in `/etc/sysctl.conf`. It would be convenient to have the changes inside docker images.
* Readme mentions about setting CPU frequency to performance, Is this important for this workload? 
* Readme could have details about `mediawiki` workload.

ajaygk95 commented:
Hello @pradeepk5 ,

Sorry for the late reply.

I have been busy in my Summer Internship.

I will start working on this PR again from Fall-2020 (Probably from Aug). Will include the suggested changes.

ajaygk95 commented:
Hello @neo-apz @pradeepk5 ,

I have updated the PR by incorporating most of the comments.

-  all images should start from our common base image (cloudsuite/debian:base-os)

    HHVM-3.30 is now built using debian:base-os and I have moved this to common folder. (This build will take around 5-6 hours and travis build are failing due to timeout). The mediawiki workload docker image now uses HHVM-3.30 image as the base image.

-  add some info on what this benchmark is about and what is trying to do (I know you have referred to the FB page on this but there should be some minimal info in the README file and then for more detailed info people can check out the original README.

    I have updated the README file with the required info

-  add multi-platform support (ARM and RISC-V if possible)

    As discussed earlier, HHVM is not available for other platforms.

---------------------------------------------------------------------------------------------------------------------------------

- perf is removed in one of the commits, but there are some remaining scripts which uses perf (perf.sh)

   I have removed all files related to perf

- seige load tester used in Django too, Can we keep move seige to common?

  siege-4.0.3 has been moved to commons folder, and the mediawiki worload now uses a remote siege.

- Readme metions about changing parameters in /etc/sysctl.conf. It would be convenient to have the changes inside docker images.

  I tried changing the `sysctl` config file inside the debian-buster docker and it was not possible. We can either set the parameters in ` /etc/sysctl.conf` or we can run the docker container with `--sysctl` flag.

- Readme mentions about setting CPU frequency to performance, Is this important for this workload?

  Yes. This is required for the workload.

- Readme could have details about mediawiki workload.

  Added info about the mediawiki workload.


Had a question, The HHVM-3.30 build needs to happen at-least once (to push to cloudsuite docker repo) and this takes around 5-6 hours. Is there any configuration related to travis build-time that we can increase ?

Thanks,
Ajay

neo-apz commented:
Thanks @ajaygk95 for the updates.
We are currently looking through the changes and will post our feedback soon.

ajaygk95 commented:
Hello,

Any updates of this ?

yunho-oh commented:
Hi @ajaygk95,
Regarding this PR, I will share the status on this Friday.
I am not aware of working on this, so I need some time to get used to the tasks relevant to this PR.
Also, I need to deal with other tasks at the same time.

Thank you for your understanding in advance.

yunho-oh commented:
Hi @ajaygk95,

I am new to this task and I may be doing something wrong. So feel free to correct me.

First, I have a question.
So far I could build the containers of the HHVM, the siege client, and the facebook workload benchmark. 
However, I cannot find and build the mysql image (or relevant sources). Where can I find it? Should it be added to the 'common' folder?
Once I can build the database server container, I will keep looking at this PR.

And regarding the 'setting scaling_governor to performance' part in fb-workload.md, I think that it would be better if we add a simple explanation of why we should do it. It would make users understand it easily.

yunho-oh commented:
I took a look at #241 now. It seems that I can try to build the MySQL after #241 is merged. I will do it then.

ajaygk95 commented:
Hello @yunho-oh ,

I guess #241 is still W.I.P @komal1204 will provide an update on that.

But you can build only the mysql docker image from #241 and test this Facebook PR and let me know your inputs.

While merging the PR's we can merge the mysql PR first and then the facebook PR.

ajaygk95 commented:
Hello @yunho-oh ,

The hhvm build is getting timed-out on the travis build.
Is there a way to increase the time-out period in travis ? 
Or can you push the hhvm:3.30 docker image to cloudsuite docker repo manually ?

komal1204 commented:
> Hello @yunho-oh ,
> 
> I guess #241 is still W.I.P @komal1204 will provide an update on that.
> 
> But you can build only the mysql docker image from #241 and test this Facebook PR and let me know your inputs.
> 
> While merging the PR's we can merge the mysql PR first and then the facebook PR.

Hello,

Yes MySQL is almost done, will be updating/creating a new pr this week.

yunho-oh commented:
Hi @ajaygk95 

I have no idea about how to enhance the time-out limit in the travis build.
I will try to push the hhvm:3.30 docker image to the docker repo and update the status no later than this Friday. Also, I will try to build the mysql container once @komal1204 updates a new PR. 
 
I have some urgent tasks as of now, so kindly ask you for understanding.

yunho-oh commented:
Hi @ajaygk95 

Regarding the hhvm image, I should get an authentication about the CloudSuite docker repository. I will get it and update the image shortly. 

Just in case, I have uploaded the hhvm image in my repository (yunhooh/hhvm:3.30).
You can use it as of now.

yunho-oh commented:
Hi, @ajaygk95 

I am doing the regression test with this PR and seeing the following error message when running this workload. I built all the images in my local server and tried to run them with the localhost IP.  Have you ever seen the following log? If I want to run all the containers on a single server, what would be the correct IP address in cmd. sh?

+ '[' '!' -f /oss-performance/cmd.sh ']'
+ bash /oss-performance/cmd.sh
+ MYSQL_IP=127.0.0.1
+ SIEGE_IP=127.0.0.1
+ mysql --host=127.0.0.1 --user=root --password=root -e 'SET GLOBAL max_connections = 1001;'
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
+ hhvm perf.php --i-am-not-benchmarking --mediawiki --db-host=127.0.0.1 --db-username=root --db-password=root --hhvm=/usr/local/bin/hhvm --remote-siege=root@127.0.0.1
ssh: connect to host 127.0.0.1 port 9801: Connection refused

Fatal error: Uncaught exception 'HH\InvariantException' with message 'Invalid ssh credentials: root@127.0.0.1' in /oss-performance/base/PerfOptions.php:396
Stack trace:
#0 /oss-performance/base/PerfOptions.php(396): HH\invariant_violation()
#1 /oss-performance/base/PerfRunner.php(48): PerfOptions->validate()
#2 /oss-performance/base/PerfRunner.php(41): PerfRunner::RunWithOptionsAndEngine()
#3 /oss-performance/base/PerfRunner.php(16): PerfRunner::RunWithOptions()
#4 /oss-performance/perf.php(14): PerfRunner::RunWithArgv()
#5 /oss-performance/perf.php(18): perf_main()
#6 {main}

ajaygk95 commented:
Hello @yunho-oh,

You can run all the containers on a single VM. Use the "private-ip" in cmd.sh.
Also, when running the docker images, make sure you are using "--net=host" flag.

ajaygk95 commented:
> Hi, @ajaygk95
> 
> I am doing the regression test with this PR and seeing the following error message when running this workload. I built all the images in my local server and tried to run them with the localhost IP. Have you ever seen the following log? If I want to run all the containers on a single server, what would be the correct IP address in cmd. sh?
> 
> * '[' '!' -f /oss-performance/cmd.sh ']'
> * bash /oss-performance/cmd.sh
> * MYSQL_IP=127.0.0.1
> * SIEGE_IP=127.0.0.1
> * mysql --host=127.0.0.1 --user=root --password=root -e 'SET GLOBAL max_connections = 1001;'
>   ERROR 1698 (28000): Access denied for user 'root'@'localhost'
> * hhvm perf.php --i-am-not-benchmarking --mediawiki --db-host=127.0.0.1 --db-username=root --db-password=root --hhvm=/usr/local/bin/hhvm --remote-siege=root@127.0.0.1
>   ssh: connect to host 127.0.0.1 port 9801: Connection refused
> 
> Fatal error: Uncaught exception 'HH\InvariantException' with message 'Invalid ssh credentials: root@127.0.0.1' in /oss-performance/base/PerfOptions.php:396
> Stack trace:
> #0 /oss-performance/base/PerfOptions.php(396): HH\invariant_violation()
> #1 /oss-performance/base/PerfRunner.php(48): PerfOptions->validate()
> #2 /oss-performance/base/PerfRunner.php(41): PerfRunner::RunWithOptionsAndEngine()
> #3 /oss-performance/base/PerfRunner.php(16): PerfRunner::RunWithOptions()
> #4 /oss-performance/perf.php(14): PerfRunner::RunWithArgv()
> #5 /oss-performance/perf.php(18): perf_main()
> #6 {main}


I guess the issue is due to the siege_ip.

If you run the siege container using "`--net=host`" flag (Make sure no service is using port 9801 on the VM), then you can use `127.0.0.1` as `SIEGE_IP` in `cmd.sh`

If you are running siege container without "`--net=host`" flag then run,
`docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' siege` to get the docker IP. And use this IP in `cmd.sh`

yunho-oh reviewed (changes requested):
I found that this workload runs on my local server. 

However, there is still an issue: I could not run this workload with the localhost. I could run it after creating a new docker network, find the IP addresses of mysql and siege, and update the IP addresses in cmd.sh manually. I think that in fb-workload.md, the explanation about 'how to add IP addresses properly' should be added so users do not have difficulty it. Could you please do it?

yunho-oh commented:
The remaining changes in this PR look ok. So, after discussing/updating what I suggested, let's quickly double-check all the changes and merge this PR.

ajaygk95 commented:
> I found that this workload runs on my local server.
> 
> However, there is still an issue: I could not run this workload with the localhost. I could run it after creating a new docker network, find the IP addresses of mysql and siege, and update the IP addresses in cmd.sh manually. I think that in fb-workload.md, the explanation about 'how to add IP addresses properly' should be added so users do not have difficulty it. Could you please do it?

The general approach to all the workloads is to not use the "docker network". Try to run the workload using "private-ip" instead of using "localhost".

Please let me know if you are still facing issues by using the private-ip

yunho-oh commented:
I missed saying that "even using private-ip", I saw the same message "Fatal error: Uncaught exception ...." is shown.

yunho-oh commented on docs/benchmarks/fb-workload.md:
````diff
@@ -0,0 +1,96 @@
+chmod +x cmd.sh
+```
+
+To run the benchmark using HHVM edit the below lines in `cmd.sh`
````
Here, I think that if we say 'edit the below lines' only, some users may misunderstand because in this manual, we suggest using '--net=host' while running this workload. If a user who already knows the web serving workload tries to run this workload, he/she might think that the IP addresses here should be the localhost IP (127.0.0.1). FYI please find [this](https://github.com/parsa-epfl/cloudsuite/blob/master/docs/benchmarks/web-serving.md), which is the readme file of the web serving workload. Honestly, I was also confused about it before you told me that I should use private ip.

So, I think that we should clarify that users should use private ip addresses in this part.

yunho-oh commented:
> I missed saying that "even using private-ip", I saw the same message [Fatal error: Uncaught exception ....] is shown.
> 
> And, I meant 'fb-workload.md' is the file in this PR. [Link](https://github.com/parsa-epfl/cloudsuite/pull/261/files#diff-3da9db09f61a292eca68a715b0fe2622b325cea0d253329b171e42d19836b922)

BTW, I will try to use another machine to test this workload using the private ip addresses. 

So, as of now, please consider only the above comment in docs/benchmarks/fb-workload.md and let me know your thought.

ajaygk95 commented:
> > I missed saying that "even using private-ip", I saw the same message [Fatal error: Uncaught exception ....] is shown.
> > And, I meant 'fb-workload.md' is the file in this PR. [Link](https://github.com/parsa-epfl/cloudsuite/pull/261/files#diff-3da9db09f61a292eca68a715b0fe2622b325cea0d253329b171e42d19836b922)
> 
> BTW, I will try to use another machine to test this workload using the private ip addresses.
> 
> So, as of now, please consider only the above comment in docs/benchmarks/fb-workload.md and let me know your thought.

Hello @yunho-oh ,

I had made the changes in readme as you suggested. Please take a look.

yunho-oh commented:
> > > I missed saying that "even using private-ip", I saw the same message [Fatal error: Uncaught exception ....] is shown.
> > > And, I meant 'fb-workload.md' is the file in this PR. [Link](https://github.com/parsa-epfl/cloudsuite/pull/261/files#diff-3da9db09f61a292eca68a715b0fe2622b325cea0d253329b171e42d19836b922)
> > 
> > 
> > BTW, I will try to use another machine to test this workload using the private ip addresses.
> > So, as of now, please consider only the above comment in docs/benchmarks/fb-workload.md and let me know your thought.
> 
> Hello @yunho-oh ,
> 
> I had made the changes in readme as you suggested. Please take a look.

Hi @ajaygk95 

Thanks for the update. It looks ok now.

I tried the regression test with another machine and I saw that this workload works with the configuration you said. 

BTW, sorry for delaying the hhvm update to the CloudSuite docker hub. Will do it asap.

Let's close this PR and merge it. If none of your team members is authorized to merge the source code, we will do it asap.

yunho-oh commented:
Hi @ajaygk95 

I missed saying one more thing: please make sure that this PR passes the travisCI to merge it asap.

ajaygk95 commented:
Hello @yunho-oh ,

All the PR builds are failing due to a common issue. I have fixed the issue in #270 . 

Please merge #270 , also is `hhvm:3.30` docker image been pushed to `cloudsuite` repo ?

yunho-oh commented:
> Hello @yunho-oh ,
> 
> All the PR builds are failing due to a common issue. I have fixed the issue in #270 .
> 
> Please review and merge #270 , also is `hhvm:3.30` docker image been pushed to `cloudsuite` repo ?

Regarding the hhvm image, yes. I have uploaded it today. Sorry for the late update. The person who is managing the docker hub gave me the authentication today.

ajaygk95 commented:
Hello @yunho-oh ,

As #270 is merged, should I rebase this branch with the `dev` branch and push the commit or will this be rebased automatically during merge ?

ajaygk95 commented:
> Hello @yunho-oh ,
> 
> As #270 is merged, should I rebase this branch with the `dev` branch and push the commit or will this be rebased automatically during merge ?

Merged
